File: showcase/integrations/langgraph-typescript/src/agent/shared-state-streaming.ts
/**
 * LangGraph TypeScript agent backing the Shared State Streaming demo.
 *
 * Demonstrates per-token state-delta streaming. The agent writes a long
 * `document` string into shared agent state via a `write_document` tool;
 * `copilotkitCustomizeConfig(..., { emitIntermediateState })` tells
 * CopilotKit to forward every token of the tool's `document` argument
 * directly into the `document` state key as it is generated. The UI
 * (useAgent) sees `state.document` grow token-by-token, without waiting
 * for the tool call to finish.
 *
 * This is the canonical per-token state-streaming pattern:
 * docs.copilotkit.ai/integrations/langgraph/shared-state/predictive-state-updates
 */

import { randomUUID } from "node:crypto";

import { z } from "zod";
import type { RunnableConfig } from "@langchain/core/runnables";
import { tool } from "@langchain/core/tools";
import { ToolNode } from "@langchain/langgraph/prebuilt";
import type { AIMessage } from "@langchain/core/messages";
import { SystemMessage, ToolMessage } from "@langchain/core/messages";
import type { ToolRunnableConfig } from "@langchain/core/tools";
import {
  Annotation,
  Command,
  MemorySaver,
  START,
  StateGraph,
} from "@langchain/langgraph";
import { ChatOpenAI } from "@langchain/openai";
import { makeChatOpenAI } from "./openai-headers";

import {
  copilotkitCustomizeConfig,
  convertActionsToDynamicStructuredTools,
  CopilotKitStateAnnotation,
} from "@copilotkit/sdk-js/langgraph";

// ---------------------------------------------------------------------------
// 1. Shared state — `document` is streamed token-by-token.
// ---------------------------------------------------------------------------

const AgentStateAnnotation = Annotation.Root({
  ...CopilotKitStateAnnotation.spec,
  document: Annotation<string>,
});

export type AgentState = typeof AgentStateAnnotation.State;

// ---------------------------------------------------------------------------
// 2. Tool — `write_document` writes the document into shared state.
// ---------------------------------------------------------------------------

const writeDocument = tool(
  async ({ document }, config: ToolRunnableConfig) => {
    const toolCallId = config.toolCall?.id;
    if (typeof toolCallId !== "string" || toolCallId.length === 0) {
      throw new Error(
        "write_document: missing tool_call_id — tool was invoked outside a " +
          "ToolNode context. Refusing to emit a ToolMessage with an empty " +
          "tool_call_id (OpenAI rejects those).",
      );
    }

    return new Command({
      update: {
        document,
        messages: [
          new ToolMessage({
            content: "Document written to shared state.",
            name: "write_document",
            id: randomUUID(),
            tool_call_id: toolCallId,
          }),
        ],
      },
    });
  },
  {
    name: "write_document",
    description:
      "Write a document for the user.\n\n" +
      "Always call this tool when the user asks you to write or draft " +
      "something of any length (an essay, poem, email, summary, etc.). " +
      "The `document` argument is streamed *per token* into shared agent " +
      "state under the `document` key, so the UI can render it as it is " +
      "generated.",
    schema: z.object({
      document: z.string(),
    }),
  },
);

const tools = [writeDocument];

// ---------------------------------------------------------------------------
// 3. Chat node.
// ---------------------------------------------------------------------------

const SYSTEM_PROMPT =
  "You are a collaborative writing assistant. Whenever the user asks " +
  "you to write, draft, or revise any piece of text, ALWAYS call the " +
  "`write_document` tool with the full content as a single string in " +
  "the `document` argument. Never paste the document into a chat " +
  "message directly — the document belongs in shared state and the " +
  "UI renders it live as you type.";

async function chatNode(state: AgentState, config: RunnableConfig) {
  const model = makeChatOpenAI(config, {
    model: "gpt-5.4",
    modelKwargs: { parallel_tool_calls: false },
  });

  const modelWithTools = model.bindTools!([
    ...convertActionsToDynamicStructuredTools(state.copilotkit?.actions ?? []),
    ...tools,
  ]);

  const systemMessage = new SystemMessage({ content: SYSTEM_PROMPT });

  // @region[state-streaming-middleware]
  const streamingConfig = copilotkitCustomizeConfig(config, {
    emitIntermediateState: [
      {
        stateKey: "document",
        tool: "write_document",
        toolArgument: "document",
      },
    ],
  });

  const response = await modelWithTools.invoke(
    [systemMessage, ...state.messages],
    streamingConfig,
  );
  // @endregion[state-streaming-middleware]

  return { messages: response };
}

// ---------------------------------------------------------------------------
// 4. Routing — send tool calls to tool_node unless they're CopilotKit
//    frontend actions.
// ---------------------------------------------------------------------------

function shouldContinue({ messages, copilotkit }: AgentState) {
  const lastMessage = messages[messages.length - 1] as AIMessage;

  if (lastMessage.tool_calls?.length) {
    const actions = copilotkit?.actions;
    const hasBackendToolCall = lastMessage.tool_calls.some((toolCall) => {
      return (
        !actions || actions.every((action) => action.name !== toolCall.name)
      );
    });

    if (hasBackendToolCall) {
      return "tool_node";
    }
  }

  return "__end__";
}

// ---------------------------------------------------------------------------
// 5. Compile the graph.
// ---------------------------------------------------------------------------

const workflow = new StateGraph(AgentStateAnnotation)
  .addNode("chat_node", chatNode)
  .addNode("tool_node", new ToolNode(tools))
  .addEdge(START, "chat_node")
  .addEdge("tool_node", "chat_node")
  .addConditionalEdges("chat_node", shouldContinue as any);

const memory = new MemorySaver();

export const graph = workflow.compile({
  checkpointer: memory,
});

File: showcase/integrations/crewai-crews/src/agents/shared_state_streaming.py
"""CrewAI Flow for predictive, per-token shared document state."""

from __future__ import annotations

import json

from crewai.flow.flow import Flow, start
from litellm import acompletion

from ag_ui_crewai import (
    CopilotKitState,
    StateItem,
    copilotkit_emit_tool_result,
    copilotkit_predict_state,
    copilotkit_stream,
)


SYSTEM_PROMPT = (
    "You are a collaborative writing assistant. Whenever the user asks you "
    "to write, draft, or revise text, always call write_document with the "
    "complete document. Do not paste the document into chat; the UI renders "
    "the tool argument live from shared state. After the tool result, reply "
    "with one short confirmation."
)
WRITE_DOCUMENT_TOOL = {
    "type": "function",
    "function": {
        "name": "write_document",
        "description": "Replace the shared document with complete new content.",
        "parameters": {
            "type": "object",
            "properties": {"document": {"type": "string"}},
            "required": ["document"],
        },
    },
}


class SharedStateStreamingState(CopilotKitState):
    document: str = ""


class SharedStateStreamingFlow(Flow[SharedStateStreamingState]):
    """Predict write_document arguments into state, then persist the result."""

    @start()
    async def write(self) -> None:
        await copilotkit_predict_state(
            [
                StateItem(
                    state_key="document",
                    tool="write_document",
                    tool_argument="document",
                )
            ]
        )
        tools = [*self.state.copilotkit.actions, WRITE_DOCUMENT_TOOL]
        response = await copilotkit_stream(
            await acompletion(
                model="openai/gpt-4o-mini",
                messages=[
                    {"role": "system", "content": SYSTEM_PROMPT},
                    *self.state.messages,
                ],
                tools=tools,
                parallel_tool_calls=False,
                stream=True,
            )
        )
        message = response.choices[0].message
        self.state.messages.append(message)
        calls = message.get("tool_calls") or []
        write_call = next(
            (
                call
                for call in calls
                if call.get("function", {}).get("name") == "write_document"
            ),
            None,
        )
        if write_call is None:
            return

        try:
            arguments = json.loads(
                write_call.get("function", {}).get("arguments") or "{}"
            )
        except (TypeError, json.JSONDecodeError):
            arguments = {}
        document = arguments.get("document")
        self.state.document = document if isinstance(document, str) else ""
        self.state.messages.append(
            {
                "role": "tool",
                "tool_call_id": write_call.get("id"),
                "content": "Document written to shared state.",
            }
        )
        await copilotkit_emit_tool_result(
            write_call.get("id"), "Document written to shared state."
        )

        confirmation = await copilotkit_stream(
            await acompletion(
                model="openai/gpt-4o-mini",
                messages=[
                    {"role": "system", "content": SYSTEM_PROMPT},
                    *self.state.messages,
                ],
                tools=tools,
                parallel_tool_calls=False,
                stream=True,
            )
        )
        self.state.messages.append(confirmation.choices[0].message)


shared_state_streaming_flow = SharedStateStreamingFlow()
